### ISSUE TITLE
Expose different types of objects in the same array

### ISSUE THREAD

**Issue by waynn** [ruby-grape/grape-entity#192]
I'd like to return a JSON array that has two different types of models, each with its own `Grape::Entity`. So for example, I have a `package`, which has this

    class Entity < Grape::Entity
      expose :id, :foo
      expose :image, format_with: :image_absolute_url
    end

And then I have `experience`, which has this

      class Entity < Grape::Entity
        expose :id, :bar
        expose :image, format_with: :image_absolute_url
      end

In my API call, I create an array like this.

    present [ Package.last, Experience.last ]

One way I thought of was to define a new entity, and pass it in using "with", but then I'm just repeating my previous definitions. And I also wasn't sure how to do it besides something like

    class NewEntity < Grape::Entity
      expose :id, if: lambda { |object, options| object.is_a?(Package) }
      expose :bar, if: lambda { |object, options| object.is_a?(Experience) }
    end

Is there a better solution for this?

**Comment by dblock** [comment]
This isn't ideal, maybe think of what you'd like to write and we can discuss how to implement it? Other than that I strongly recommend not doing this and refactoring your entities with OO so that you don't have to do IFs here.

**Comment by waynn** [comment]
Yeah, it's definitely clunky. So I'm trying to build a home feed for an application, but the content is going to be mixed and composed of these two types (like Yelp mixing categories and specific restaurants, or Airbnb mixing cities and specific houses). I wanted to use an array because there's a defined order to the feed. One other way is I could add a new module to both classes that implements some subset of the functionality, so that the logic gets abstracted there instead of inside the entity, though it probably makes sense to just keep it in in the entity.

Is it possible to put all that logic inside the entity? I couldn't access the instance outside of the lambda block.

Or any other suggestions more than welcome. :)

**Comment by dblock** [comment]
I would build this with a HAL API today (and would use grape + roar). That layout allows you to combine different types of things under `_embeded` in a way that doesn't mix apples and oranges.

**Comment by waynn** [comment]
Ah, ok. One more quick follow up, that might make things a little easier. Is there a way to present different attributes all at once, instead of with an if: clause on each attribute? Something like

    if object.is_a?(Category)
      expose :id, :name, :description, :foo
    else
      expose :bar, :baz
    end

**Comment by dblock** [comment]
Not ATM. IMHO this should not be a requirement.

**Comment by DolevG** [comment]
google protobuff added the one of feature, https://developers.google.com/protocol-buffers/docs/proto#oneof . it's relatively easy to manipulate grape to return that kind of proto

**Comment by dblock** [comment]
> > I would build this with a HAL or JSON+API today (and would use grape + roar)
> 
> Could someone expand on this comment a little bit? 

I was saying that I wouldn't choose Grape::Entity if I wanted something like this.
